mm2src/coins/eth/eip1559_gas_fee.rs

```diff
@@ -0,0 +1,469 @@
+                    .clone(),
+                min_wait_time: None,
+                max_wait_time: None,
+            },
+            source: EstimationSource::Blocknative,
+            units: EstimationUnits::Gwei,
+            base_fee_trend: String::default(),
+            priority_fee_trend: String::default(),
+        }
+    }
+}
+
+/// Simple priority fee per gas estimator based on fee history
+/// normally used if gas api provider is not available
+pub(super) struct FeePerGasSimpleEstimator {}
+
+impl FeePerGasSimpleEstimator {
+    // TODO: add minimal max fee and priority fee
+    /// depth to look for fee history to estimate priority fees
+    const FEE_PRIORITY_DEPTH: u64 = 5u64;
+
+    /// percentiles to pass to eth_feeHistory
+    const HISTORY_PERCENTILES: [f64; FEE_PER_GAS_LEVELS] = [25.0, 50.0, 75.0];
+
+    /// percentiles to calc max priority fee over historical rewards
+    const CALC_PERCENTILES: [f64; FEE_PER_GAS_LEVELS] = [50.0, 50.0, 50.0];
+
+    /// adjustment for max priority fee picked up by sampling
+    const ADJUST_MAX_FEE: [f64; FEE_PER_GAS_LEVELS] = [1.0, 1.0, 1.0];
+
+    /// adjustment for max priority fee picked up by sampling
+    const ADJUST_MAX_PRIORITY_FEE: [f64; FEE_PER_GAS_LEVELS] = [1.0, 1.0, 1.0];
+
+    /// block depth for eth_feeHistory
+    pub fn history_depth() -> u64 { Self::FEE_PRIORITY_DEPTH }
+
+    /// percentiles for priority rewards obtained with eth_feeHistory
+    pub fn history_percentiles() -> &'static [f64] { &Self::HISTORY_PERCENTILES }
+
+    /// percentile for vector
+    fn percentile_of(v: &mut Vec<U256>, percent: f64) -> U256 {
+        v.sort();
+
+        // validate bounds:
+        let percent = if percent > 100.0 { 100.0 } else { percent };
+        let percent = if percent < 0.0 { 0.0 } else { percent };
+
+        let value_pos = ((v.len() - 1) as f64 * percent / 100.0).round() as usize;
+        v[value_pos]
+    }
+
+    /// Estimate simplified gas priority fees based on fee history
+    pub async fn estimate_fee_by_history<T: Transport>(
+        fee_history_namespace: EthFeeHistoryNamespace<T>,
+    ) -> Web3RpcResult<FeePerGasEstimated> {
+        let res = fee_history_namespace
+            .eth_fee_history(
+                U256::from(Self::history_depth()),
+                BlockNumber::Latest,
+                Self::history_percentiles(),
+            )
+            .await;
+
+        match res {
+            Ok(fee_history) => Ok(Self::calculate_with_history(&fee_history)),
+            Err(_) => MmError::err(Web3RpcError::Internal("Eth requests failed".into())),
+        }
+    }
+
+    /// estimate priority fees by fee history
+    fn calculate_with_history(fee_history: &FeeHistoryResult) -> FeePerGasEstimated {
+        let base_fee = *fee_history.base_fee_per_gas.first().unwrap_or(&U256::from(0));
+        let base_fee = u256_to_big_decimal(base_fee, ETH_GWEI_DECIMALS).unwrap_or_else(|_| BigDecimal::from(0));
+        let mut priority_fees = vec![];
```

Comment:
Using `vec![]` is overhead here, as it is always supposed to have `len == 3`. It can be a fixed-size array, pre-initialized with default values.

Reply:
refactored this code